app/src/main/java/org/wikipedia/diff/ArticleEditDetailsFragment.kt

```diff
@@ -249,12 +273,43 @@ class ArticleEditDetailsFragment : Fragment(), WatchlistExpiryDialog.Callback, L
         }
 
         binding.undoButton.setOnClickListener {
-            showUndoDialog()
-            editHistoryInteractionEvent?.logUndoTry()
+            val canUndo = viewModel.revisionFrom != null && AccountUtil.isLoggedIn
+            val canRollback = AccountUtil.isLoggedIn && viewModel.hasRollbackRights && !viewModel.canGoForward
+
+            if (canUndo && canRollback) {
+                PopupMenu(requireContext(), binding.undoLabel, Gravity.END).apply {
+                    menuInflater.inflate(R.menu.menu_context_undo, menu)
+                    setForceShowIcon(true)
+                    setOnMenuItemClickListener { menuItem ->
+                        when (menuItem.itemId) {
+                            R.id.menu_undo -> {
+                                showUndoDialog()
+                                editHistoryInteractionEvent?.logUndoTry()
+                                true
+                            }
+                            R.id.menu_rollback -> {
+                                showRollbackDialog()
+                                true
+                            }
+                            else -> false
+                        }
+                    }
+                    show()
+                }
+            } else if (canUndo) {
+                showUndoDialog()
+                editHistoryInteractionEvent?.logUndoTry()
+            }
+        }
+
+        binding.watchButton.setOnClickListener {
+            viewModel.watchOrUnwatch(isWatched, WatchlistExpiry.NEVER, isWatched)
+            if (isWatched) editHistoryInteractionEvent?.logUnwatchClick() else editHistoryInteractionEvent?.logWatchClick()
         }
+        updateWatchButton(isWatched, hasWatchlistExpiry)
 
-        binding.rollbackButton.setOnClickListener {
-            showRollbackDialog()
+        binding.warnButton.setOnClickListener {
+            // TODO: implement this
```

Comment:
Could we please connect the warn flow here?

Reply:
This will be implemented in the PR of connecting warn template with the reply activity.